pilot/pkg/security/authz/builder/extauthz.go

```diff
@@ -0,0 +1,317 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package builder
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/davecgh/go-spew/spew"
+	envoy_config_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	extauthzhttp "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/ext_authz/v3"
+	extauthztcp "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/ext_authz/v3"
+	envoy_type_matcher_v3 "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+	envoytypev3 "github.com/envoyproxy/go-control-plane/envoy/type/v3"
+	"github.com/golang/protobuf/ptypes/duration"
+	"github.com/hashicorp/go-multierror"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	authzmodel "istio.io/istio/pilot/pkg/security/authz/model"
+	"istio.io/istio/pkg/config/host"
+)
+
+const (
+	extAuthzMatchPrefix = "istio-ext-authz"
+)
+
+var (
+	rbacDefaultDenyAll = &rbacpb.RBAC{
+		Action: rbacpb.RBAC_DENY,
+		Policies: map[string]*rbacpb.Policy{
+			"default-deny-all-due-to-bad-CUSTOM-action": {
+				Permissions: []*rbacpb.Permission{{Rule: &rbacpb.Permission_Any{Any: true}}},
+				Principals:  []*rbacpb.Principal{{Identifier: &rbacpb.Principal_Any{Any: true}}},
+			},
+		},
+	}
+)
+
+type builtExtAuthz struct {
+	http *extauthzhttp.ExtAuthz
+	tcp  *extauthztcp.ExtAuthz
+}
+
+func notAllTheSame(names []string) bool {
+	for i := 1; i < len(names); i++ {
+		if names[i-1] != names[i] {
+			return true
+		}
+	}
+	return false
+}
+
+func buildExtAuthz(configs []*meshconfig.MeshConfig_ExtensionProvider, providers []string) (*builtExtAuthz, error) {
+	if notAllTheSame(providers) {
+		return nil, fmt.Errorf("all extension providers must be the same for a specific workload, found multiple different providers: %v", providers)
+	} else if len(providers) < 1 {
+		return nil, fmt.Errorf("no extension provider found")
+	}
+	provider := providers[0]
+
+	resolved := map[string]*builtExtAuthz{}
+	var errs error
+	if len(configs) == 0 {
+		errs = multierror.Append(errs, fmt.Errorf("at least 1 extension provider must be defined"))
+	}
+	for i, config := range configs {
+		if config.Name == "" {
+			errs = multierror.Append(errs, fmt.Errorf("extension provider name must not be empty, found empty at index: %d", i))
+		} else if _, found := resolved[config.Name]; found {
+			errs = multierror.Append(errs, fmt.Errorf("extension provider name must be unique, found duplicate: %s", config.Name))
```

Comment:
This should be validated during mesh configuration loading

Reply:
how? This needs the list of authz policies applied for a given workload, we don't know the workload in mesh configuration loading. Also I'm not sure how useful this could be, it seems the final result is mostly the same regardless where you do the validation.

Reply:
oh, I looked at the wrong line (70), but actually this is also validated in mesh configuration loading in pkg/config/validation/extensionprovider.go

Reply:
I am still confused about why we need to duplicate validating it.

Reply:
there are some validation that we can only do here because we don't know the workload and policies applied to it in ValidateMeshConfig().

Reply:
make sense